src/twisted/protocols/_smb/_base.py

```diff
@@ -0,0 +1,396 @@
+        type=int,
+        metadata={SMB_METADATA: "I"},
+        validator=default_only if locked else None,
+    )
+
+
+def long(default: int = 0, locked: bool = False) -> Any:
+    """an 64-bit unsigned integer"""
+    return attr.ib(
+        default=default,
+        type=int,
+        metadata={SMB_METADATA: "Q"},
+        validator=default_only if locked else None,
+    )
+
+
+def single(default: float = 0.0) -> Any:
+    """a 32-bit float"""
+    return attr.ib(default=default, type=float, metadata={SMB_METADATA: "f"})
+
+
+def double(default: float = 0.0) -> Any:
+    """a 64-bit float"""
+    return attr.ib(default=default, type=float, metadata={SMB_METADATA: "d"})
+
+
+def octets(
+    length: Optional[int] = None, default: Optional[bytes] = None, locked: bool = False
+) -> Any:
+    """
+    a group of octets (bytes). Either a length or a default must be given.
+    If a length, the default is all zeros, if a default, the length is taken
+    from the default.
+
+    @param length: number of bytes
+    @type length: L{int}
+
+    @type default: L{bytes}
+    """
+    assert length or default
+    if length is None:
+        length = len(cast(bytes, default))
+    if default is None:
+        default = b"\0" * length
+    return attr.ib(
+        default=default,
+        type=bytes,
+        metadata={SMB_METADATA: str(length) + "s"},
+        validator=default_only if locked else None,
+    )
+
+
+NULL_UUID = uuid_mod.UUID("00000000-0000-0000-0000-000000000000")
+NEW_UUID = attr.Factory(uuid_mod.uuid4)
+
+
+def uuid(default: uuid_mod.UUID = NULL_UUID, locked: bool = False) -> Any:
+    """a universial unique ID"""
+    default = _conv_uuid(default)
+    return attr.ib(
+        default=default,
+        metadata={SMB_METADATA: "16s"},
+        type=uuid_mod.UUID,
+        converter=_conv_uuid,
+        validator=default_only if locked else None,
+    )
+
+
+def _conv_uuid(x: Union[str, bytes, uuid_mod.UUID]) -> uuid_mod.UUID:
+    if type(x) is str:
+        return uuid_mod.UUID(x)
+    elif type(x) is bytes:
+        return uuid_mod.UUID(bytes_le=x)
+    else:
+        return cast(uuid_mod.UUID, x)
+
+
+def pack(obj: Any) -> bytes:
+    """
+    pack an object into binary data. The object must have been decorated
+    with L{attr.s} and the fields set with the appropriate metadata using
+    the helpers in this module
+
+    @rtype: L{bytes}
+    """
+    strct = _get_struct(type(obj))
+    args = tuple(_conv_arg(obj, i) for i in smb_fields(type(obj)))
+    log.debug("about to pack {strct!r} {args}", strct=strct, args=args)
+    return strct.pack(*args)
+
+
+def _conv_arg(obj: Any, attrib: attr.Attribute) -> Any:
+    val = getattr(obj, attrib.name)
+    if type(val) is uuid_mod.UUID:
+        val = val.bytes_le
+    if isinstance(val, enum.Enum):
+        val = val.value
+    return val
+
+
+IGNORE = 0
+ERROR = 1
+OFFSET = 2
+DATA = 3
+
+
+def unpack(cls: Any, data: bytes, offset: int = 0, remainder: int = IGNORE) -> Any:
```

Comment:
There is a way to properly express "takes a type, returns an instance of that type" with mypy; you don't have to use `Any` here:

```suggestion
T = TypeVar("T")
def unpack(cls: Type[T], data: bytes, offset: int = 0, remainder: int = IGNORE) -> T:
```